tangelo/toolboxes/molecular_computation/fno.py

```diff
@@ -0,0 +1,415 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Module containing datastructures for interfacing with the Frozen Natural
+Orbitals (FNOs), to automatically truncate the virtual orbital space.
+
+Reference: arXiv:2002.07901
+"""
+
+from typing import List, Sequence, Tuple, Union
+import warnings
+
+import numpy as np
+
+from tangelo.algorithms.classical import MP2Solver
+from tangelo.toolboxes.molecular_computation.molecule import SecondQuantizedMolecule
+
+
+class FNO():
+    """Class to interface with the Frozen Natural Orbitals protocol, that aims
+    at reducing the computational cost of a molecular problem by truncating
+    the virtual space. In general, the virtual orbitals are ranked according to
+    their MP2 occupancies, and selected with a given threshold. They are also
+    transformed to the FNO basis, using the eigenvectors of the MP2
+    virtual-virtual density matrix.
+
+    Attributes:
+            sec_mol (SecondQuantizedMolecule): Self-explanatory.
+            uhf (bool): Flag indicating the type of mean field used.
+            n_mos (int): Number of molecular orbitals.
+            fock_ao (np.array): Fock matrix in atomic orbital form.
+            frozen_occupied (list): List of indices of frozen occupied orbitals.
+            threshold (float or list): Threshold(s) for FNO occupancy.
+
+        Properties:
+            updated_sec_mol: Returns an updated SecondQuantizedMolecule object with
+                frozen orbitals and updated MO coefficients.
+
+        Methods:
+            compute_fno: Computes and truncates the FNO orbitals.
+            get_frozen_indices: Determines the indices of the frozen orbitals.
+            diagonalize_and_reorder: Diagonalizes a matrix and can reorder the
+                eigenvalues and eigenvectors based on occupations.
+            get_number_of_fnos_from_frac_occupancies: Calculates the number of
+                active virtual orbitals based on fractional occupancies.
+    """
+
+    def __init__(self, sec_mol: SecondQuantizedMolecule, threshold: Union[float, Sequence[float]]):
+        """Initialization of the FNO class instance.
+
+        Checks for frozen virtual orbitals and warns if they are set, as they
+        might be overwritten.
+
+        Args:
+            sec_mol (SecondQuantizedMolecule): The SecondQuantizedMolecule
+                object.
+            threshold (float or list): Threshold(s) for FNO occupancy.
+        """
+
+        # Check if there are frozen virtual orbitals. Print warning that this
+        # setting will be overwritten.
+        if sec_mol.frozen_virtual and sec_mol.frozen_virtual != [[], []]:
+            warnings.warn(f"The frozen orbitals {sec_mol.frozen_virtual} indices are likely to be overwritten in {self.__class__.__name__}.", RuntimeWarning)
```

Comment:
"likely" ? If you mean that for sure we are writing over these values (whether or not the new values are the same of different) I think we should just say we overwrite them.